tutorials/learn_digital_audio/02_BasicOsc.cpp

```diff
@@ -0,0 +1,115 @@
+// Joel A. Jaffe 2024-07-17
+
+/* Basic Oscillator
+This app demonstrates a basic phase accumulator class 
+*/
+
+#ifndef MAIN // because main() cannot be overridden,
+#define MAIN 2  // we use 'preprocessor macros' 
+#endif // to conditionally compile main() 
+#include "01_Intro.cpp" // in apps that #include others
+
+/**
+ * @brief Basic phase accumulator. 
+ * Produces an ideal unipolar sawtooth wave given a sample rate and frequency. 
+ */
+template <typename T>
+class Phasor {
+private:
+int sampleRate;
+  T phase = 0; // the value an oscillator tracks over time
+  T frequency = 1; // determines how often the waveform repeats
+  T phaseIncrement = frequency / sampleRate; // determines how much to increment phase by every frame
+
+public:
+  Phasor(int sampleRate) : sampleRate(sampleRate) { // constructor 
+    phaseIncrement = frequency / sampleRate;
+  }
+  virtual ~Phasor() {} // destructor 
+
+  /**
+   * @brief Sets the oscillator's frequency
+   * @param freqHz frequency in hertz (cycles per second)
+   */
+  virtual void setFrequency(T freqHz) { // notice that this function is 'virtual', meaning it can 
+    frequency = freqHz;                 // be overridden in classes that inherit from this one
+    phaseIncrement = ::abs(this->frequency) / static_cast<T>(this->sampleRate);
+  }
+
+  /**
+   * @brief Increments and returns `phase` 
+   * @return `phase` (after increment)
+   */
+  virtual T processSample() {
+    phase += phaseIncrement; // increment phase
+    if (phase >= 1) {phase -= 1;} // if waveform zenith, wrap phase
```

Comment:
Usually you have a while loop here to guard against phase increments larger than 1. Perhaps better, you should clamp the increment in setFrequency. Also, usually negative frequencies are permitted which may be required for FM and can be quite useful for LFOs.

Reply:
Yes, just do the range check in setFrequency. You may also want to use phase - floor(phase) to do the wrap instead of the conditional. I would say it's less error prone and probably just as efficient. (Note that std::fmod would seem like an alternative, however, it can return negative values.) Another possible issue here is that you pre-inc the phase. This means if you set the osc phase and then request the next sample, it will always be one sample ahead. You can either cache the value before incrementing or subtract the phaseInc from the phase when setting the phase.